plugin/src/main/java/io/fabric8/maven/plugin/BuildMojo.java

```diff
@@ -241,7 +293,191 @@ private void executeOpenShiftBuild(ServiceHub hub, ImageConfiguration imageConfi
+
+        waitForOpenShiftBuildToComplete(client, buildName, build);
+
+        updateImageStreamTags(client, imageConfig, buildName, build);
+    }
+
+
+    private void waitForOpenShiftBuildToComplete(OpenShiftClient client, String buildConfigName, Build build) {
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        String buildName = getName(build);
+        Watcher<Build> buildWatcher = new Watcher<Build>() {
+            @Override
+            public void eventReceived(Action action, Build resource) {
+                if (isBuildCompleted(action, resource)) {
+                    latch.countDown();
+                }
+            }
+
+            @Override
+            public void onClose(KubernetesClientException cause) {
+            }
+        };
+        log.info("Waiting for build " + buildName + " to complete...");
+        try (Watch watcher = client.builds().withName(buildName).watch(buildWatcher)) {
+            while (latch.getCount() > 0L) {
+                try {
+                    latch.await();
+                } catch (InterruptedException e) {
+                    // ignore
+                }
+            }
+            log.info("Build " + buildName + " completed!");
+        }
+    }
+
+
+    /**
+     * Lets update the ImageStream to include the correct tags
+     */
+    private void updateImageStreamTags(OpenShiftClient client, ImageConfiguration imageConfig, String buildConfigName, Build build) throws MojoExecutionException {
+
+        try {
+            File manifest = openshiftManifest;
+            if (!Files.isFile(manifest)) {
+                throw new MojoFailureException("No such generated manifest file: " + manifest);
+            }
+
+            String namespace = clusterAccess.getNamespace();
+            Controller controller = new Controller(client);
+
+            boolean updated = false;
+            Set<HasMetadata> entities = loadResources(client, controller, namespace, manifest, project, log);
+            for (HasMetadata entity : entities) {
+                if (entity instanceof ImageStream) {
+                    ImageStream is = (ImageStream) entity;
+                    String imageStreamName = KubernetesHelper.getName(is);
+                    if (Objects.equals(buildConfigName, imageStreamName)) {
+                        if (updateImageStreamTag(client, imageConfig, is, buildConfigName)) {
+                            updated = true;
+                        }
+                    }
+                }
+            }
+            if (updated) {
+                // lets store the entities again!
+                KubernetesList entity = new KubernetesListBuilder().withItems(new ArrayList<>(entities)).build();
+                File resourceFileBase = new File(this.targetDir, ResourceClassifier.OPENSHIFT.getValue());
+                AbstractResourceMojo.writeResourcesIndividualAndComposite(entity, resourceFileBase, ResourceFileType.yaml, log);
```

Comment:
I see you point, but there are also other reasons which might speak against this:

* One reason why we put the `ResourceMojo` before the `BuildMojo` was that the generated resource descriptors are part of the jar file. If now the `BuildMojo` is required to update the descriptors *after* the build, the included descriptors are not valid anymore.

* Another thing is, isn't the `ImageStreamTag` *installation* specific ? So, does it make sense to distribute these in released artifacts at all (if e.g. the imagestream tag would be different on the production side). From my understanding, the `ResourceMojo` creates installation independent resource descriptors, which are universally applicable (and not only for the particular setup). These are included in the artifact jar and image itself. The `BuildMojo` creates images or other build related objects. The `DeployMojo` applies the universal resources and may add and mangle installation specific objects like `ImageStreamTags`.

Point 1 is a bit an egg-hen situation. I have no problem to give up on distributing the resource descriptors as part of the jar (we would need to change quite a bit for the devops stuff of how we combine descriptors to larger apps, though). 

Another idea would be to distinguish the resource descriptors between 'installation independent' (these are then released and distributed over e.g. a Maven repo) and 'installation dependen' (these we would keep locally)

imo its not so much about speed (its also important, of course), but how and what we deliver as part of the artifacts.